File: examples/llm.zig
//! LLM Inference Example
//!
//! Demonstrates local LLM inference with:
//! - GGUF model loading
//! - Tokenization (BPE/SentencePiece)
//! - Text generation with sampling
//! - Streaming output

const std = @import("std");
const abi = @import("abi");

pub fn main(init: std.process.Init.Minimal) !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    std.debug.print("=== ABI LLM Inference Example ===\n\n", .{});

    if (!abi.ai.isEnabled()) {
        std.debug.print("AI feature is disabled. Enable with -Denable-ai=true\n", .{});
        return;
    }

    // Initialize framework
    var framework = abi.initWithConfig(allocator, .{
        .ai = .{ .llm = .{} },
    }) catch |err| {
        std.debug.print("Framework initialization failed: {}\n", .{err});
        return err;
    };
    defer framework.deinit();

    // === Model Loading Demo ===
    std.debug.print("--- Model Loading ---\n", .{});

    // Check for model file argument or use default
    const model_path = blk: {
        var args_it = init.args.iterateAllocator(allocator) catch |err| {
            std.debug.print("Failed to read args: {t}\n", .{err});
            break :blk "model.gguf";
        };
        defer args_it.deinit();
        _ = args_it.next(); // Skip executable name.
        if (args_it.next()) |arg| break :blk arg[0..arg.len];
        break :blk "model.gguf";
    };

    std.debug.print("Looking for model: {s}\n", .{model_path});

    // Try to load model (will fail gracefully if not found)
    var model = abi.ai.llm.Model.load(allocator, model_path) catch |err| {
        std.debug.print("\nModel not found or failed to load: {}\n", .{err});
        std.debug.print("\nTo use this example:\n", .{});
        std.debug.print("  1. Download a GGUF model (e.g., from HuggingFace)\n", .{});
        std.debug.print("  2. Run: zig build run-llm -- path/to/model.gguf\n", .{});
        std.debug.print("\nShowing API demo without model...\n\n", .{});

        // Demo the API structure without actual model
        demoApiStructure();
        return;
    };
    defer model.deinit();

    // === Model Info ===
    std.debug.print("\n--- Model Information ---\n", .{});
    const info = model.info();
    std.debug.print("Architecture: {s}\n", .{info.architecture});
    std.debug.print("Hidden dim: {d}\n", .{info.dim});
    std.debug.print("Layers: {d}\n", .{info.n_layers});
    std.debug.print("Heads: {d} (KV: {d})\n", .{ info.n_heads, info.n_kv_heads });
    std.debug.print("Vocab size: {d}\n", .{info.vocab_size});
    std.debug.print("Max context: {d}\n", .{info.max_seq_len});
    std.debug.print("KV cache: {B}\n", .{info.kv_cache_memory});
    std.debug.print("Weights: {B}\n", .{info.weights_memory});

    // === Tokenization Demo ===
    std.debug.print("\n--- Tokenization ---\n", .{});
    const test_text = "Hello, world! How are you today?";
    const tokens = model.encode(test_text) catch |err| {
        std.debug.print("Tokenization failed: {}\n", .{err});
        return err;
    };
    defer allocator.free(tokens);

    std.debug.print("Input: \"{s}\"\n", .{test_text});
    std.debug.print("Tokens ({d}): ", .{tokens.len});
    for (tokens) |tok| {
        std.debug.print("{d} ", .{tok});
    }
    std.debug.print("\n", .{});

    // Decode back
    const decoded = model.decode(tokens) catch |err| {
        std.debug.print("Detokenization failed: {}\n", .{err});
        return err;
    };
    defer allocator.free(decoded);
    std.debug.print("Decoded: \"{s}\"\n", .{decoded});

    // === Generation Demo ===
    std.debug.print("\n--- Text Generation ---\n", .{});

    const prompt = "Once upon a time";
    std.debug.print("Prompt: \"{s}\"\n", .{prompt});
    std.debug.print("Generating...\n\n", .{});

    // Configure generation
    const gen_config = abi.ai.llm.generation.GeneratorConfig{
        .temperature = 0.8,
        .top_k = 40,
        .top_p = 0.95,
        .repetition_penalty = 1.1,
        .max_tokens = 50,
    };

    const prompt_tokens = model.encode(prompt) catch |err| {
        std.debug.print("Prompt encoding failed: {}\n", .{err});
        return err;
    };
    defer allocator.free(prompt_tokens);

    var timer = std.time.Timer.start() catch null;
    const output_tokens = model.generate(prompt_tokens, gen_config) catch |err| {
        std.debug.print("Generation failed: {}\n", .{err});
        return err;
    };
    const elapsed_ns = if (timer) |*t| t.read() else 0;
    defer allocator.free(output_tokens);

    const output_text = model.decode(output_tokens) catch |err| {
        std.debug.print("Decoding failed: {}\n", .{err});
        return err;
    };
    defer allocator.free(output_text);

    std.debug.print("{s}{s}\n\n", .{ prompt, output_text });

    // === Generation Stats ===
    std.debug.print("--- Generation Stats ---\n", .{});
    const token_count = output_tokens.len;
    std.debug.print("Tokens generated: {d}\n", .{token_count});
    if (elapsed_ns > 0) {
        const tokens_per_second = @as(f64, @floatFromInt(token_count)) /
            (@as(f64, @floatFromInt(elapsed_ns)) / 1_000_000_000.0);
        std.debug.print("Tokens/second: {d:.2}\n", .{tokens_per_second});
        std.debug.print("Generation time: {d:.2}ms\n", .{@as(f64, @floatFromInt(elapsed_ns)) / 1_000_000.0});
    }

    std.debug.print("\n=== LLM Example Complete ===\n", .{});
}

/// Demo API structure when model is not available
fn demoApiStructure() void {
    std.debug.print("--- LLM API Structure Demo ---\n\n", .{});

    std.debug.print("1. Load Model:\n", .{});
    std.debug.print("   var model = try abi.ai.llm.Model.load(allocator, \"model.gguf\");\n", .{});
    std.debug.print("   defer model.deinit();\n\n", .{});

    std.debug.print("2. Tokenize Text:\n", .{});
    std.debug.print("   const tokens = try model.encode(\"Hello world\");\n", .{});
    std.debug.print("   defer allocator.free(tokens);\n\n", .{});

    std.debug.print("3. Configure Sampling:\n", .{});
    std.debug.print("   const config = abi.ai.llm.generation.GeneratorConfig{{\n", .{});
    std.debug.print("       .temperature = 0.8,\n", .{});
    std.debug.print("       .top_k = 40,\n", .{});
    std.debug.print("       .top_p = 0.95,\n", .{});
    std.debug.print("   }};\n\n", .{});

    std.debug.print("4. Generate Text:\n", .{});
    std.debug.print("   const prompt_tokens = try model.encode(\"Once upon a time\");\n", .{});
    std.debug.print("   defer allocator.free(prompt_tokens);\n", .{});
    std.debug.print("   const output_tokens = try model.generate(prompt_tokens, config);\n", .{});
    std.debug.print("   defer allocator.free(output_tokens);\n", .{});
    std.debug.print("   const output = try model.decode(output_tokens);\n", .{});
    std.debug.print("   defer allocator.free(output);\n", .{});
    std.debug.print("   std.debug.print(\"Once upon a time{{s}}\", .{{output}});\n\n", .{});

    std.debug.print("5. CLI Commands:\n", .{});
    std.debug.print("   zig build run -- llm info model.gguf\n", .{});
    std.debug.print("   zig build run -- llm generate model.gguf --prompt \"Hello\"\n", .{});
    std.debug.print("   zig build run -- llm chat model.gguf\n", .{});
    std.debug.print("   zig build run -- llm bench model.gguf\n\n", .{});

    std.debug.print("Supported formats: GGUF (Q4_0, Q4_1, Q5_0, Q5_1, Q8_0)\n", .{});
    std.debug.print("Supported tokenizers: BPE, SentencePiece\n", .{});
}

File: bindings/python/abi/llm.py
"""
ABI Framework LLM Module

Provides local LLM inference capabilities supporting GGUF models
and transformer architectures.
"""

from typing import List, Optional, Dict, Any, Callable, Iterator
from dataclasses import dataclass, field
from enum import Enum, auto
import os


class SamplingMethod(Enum):
    """Token sampling methods."""
    GREEDY = auto()
    TOP_K = auto()
    TOP_P = auto()
    TEMPERATURE = auto()
    MIROSTAT = auto()
    MIROSTAT_V2 = auto()


@dataclass
class InferenceConfig:
    """
    Configuration for LLM inference.

    Attributes:
        max_context_length: Maximum context length in tokens
        max_new_tokens: Maximum tokens to generate
        temperature: Sampling temperature (0.0 = greedy, 1.0 = default)
        top_p: Top-p nucleus sampling threshold
        top_k: Top-k sampling (0 = disabled)
        repetition_penalty: Repetition penalty (1.0 = disabled)
        use_gpu: Use GPU acceleration if available
        num_threads: Number of CPU threads (0 = auto-detect)
        streaming: Enable streaming output
        batch_size: Batch size for prefill

    Example:
        >>> config = InferenceConfig(temperature=0.7, top_p=0.9)
        >>> config = InferenceConfig(max_new_tokens=512, use_gpu=True)
    """
    max_context_length: int = 2048
    max_new_tokens: int = 256
    temperature: float = 0.7
    top_p: float = 0.9
    top_k: int = 40
    repetition_penalty: float = 1.1
    use_gpu: bool = True
    num_threads: int = 0
    streaming: bool = True
    batch_size: int = 512

    @classmethod
    def default(cls) -> "InferenceConfig":
        """Create default inference configuration."""
        return cls()

    @classmethod
    def greedy(cls) -> "InferenceConfig":
        """Create greedy sampling configuration."""
        return cls(temperature=0.0, top_k=1)

    @classmethod
    def creative(cls) -> "InferenceConfig":
        """Create creative sampling configuration."""
        return cls(temperature=1.0, top_p=0.95, top_k=100)


@dataclass
class InferenceStats:
    """
    Statistics from LLM inference.

    Attributes:
        load_time_ns: Time to load model in nanoseconds
        prefill_time_ns: Time for prompt processing in nanoseconds
        decode_time_ns: Time for token generation in nanoseconds
        prompt_tokens: Number of prompt tokens processed
        generated_tokens: Number of tokens generated
        peak_memory_bytes: Peak memory usage in bytes
        used_gpu: Whether GPU was used
    """
    load_time_ns: int = 0
    prefill_time_ns: int = 0
    decode_time_ns: int = 0
    prompt_tokens: int = 0
    generated_tokens: int = 0
    peak_memory_bytes: int = 0
    used_gpu: bool = False

    @property
    def prefill_tokens_per_second(self) -> float:
        """Calculate prefill throughput."""
        if self.prefill_time_ns == 0:
            return 0.0
        return self.prompt_tokens / (self.prefill_time_ns / 1_000_000_000)

    @property
    def decode_tokens_per_second(self) -> float:
        """Calculate decode throughput."""
        if self.decode_time_ns == 0:
            return 0.0
        return self.generated_tokens / (self.decode_time_ns / 1_000_000_000)

    @property
    def total_time_seconds(self) -> float:
        """Get total inference time in seconds."""
        return (self.prefill_time_ns + self.decode_time_ns) / 1_000_000_000

    def __repr__(self) -> str:
        return (
            f"InferenceStats(prefill={self.prefill_tokens_per_second:.1f} tok/s, "
            f"decode={self.decode_tokens_per_second:.1f} tok/s, "
            f"prompt={self.prompt_tokens}, generated={self.generated_tokens}, "
            f"gpu={self.used_gpu})"
        )


@dataclass
class ModelInfo:
    """
    Information about a loaded model.

    Attributes:
        name: Model name
        path: Model file path
        architecture: Model architecture (e.g., "llama", "mistral")
        parameters: Number of parameters
        context_length: Maximum context length
        vocab_size: Vocabulary size
        quantization: Quantization type (e.g., "Q4_0", "Q8_0")
        file_size_bytes: Model file size in bytes
    """
    name: str = ""
    path: str = ""
    architecture: str = ""
    parameters: int = 0
    context_length: int = 0
    vocab_size: int = 0
    quantization: str = ""
    file_size_bytes: int = 0


class LlmEngine:
    """
    High-level interface for loading and running LLM models.

    This class provides a pure Python implementation with optional
    native acceleration when the ABI library is available.

    Example:
        >>> engine = LlmEngine()
        >>> engine.load_model("./models/llama-7b.gguf")
        >>> response = engine.generate("Hello, how are you?")
        >>> print(response)

        >>> # Streaming generation
        >>> for token in engine.generate_streaming("Once upon a time"):
        ...     print(token, end="", flush=True)
    """

    def __init__(self, config: Optional[InferenceConfig] = None):
        """
        Initialize the LLM engine.

        Args:
            config: Inference configuration (uses defaults if not provided)
        """
        self._config = config or InferenceConfig.default()
        self._model: Optional[Dict[str, Any]] = None
        self._model_info: Optional[ModelInfo] = None
        self._stats = InferenceStats()
        self._lib = None

        # Try to load native library
        try:
            from . import _load_library
            self._lib = _load_library()
        except (ImportError, AttributeError):
            pass

    @property
    def config(self) -> InferenceConfig:
        """Get current inference configuration."""
        return self._config

    @config.setter
    def config(self, value: InferenceConfig) -> None:
        """Set inference configuration."""
        self._config = value

    @property
    def model_info(self) -> Optional[ModelInfo]:
        """Get information about the loaded model."""
        return self._model_info

    @property
    def stats(self) -> InferenceStats:
        """Get inference statistics."""
        return self._stats

    @property
    def is_loaded(self) -> bool:
        """Check if a model is loaded."""
        return self._model is not None

    def load_model(self, path: str) -> ModelInfo:
        """
        Load a model from a GGUF file.

        Args:
            path: Path to the GGUF model file

        Returns:
            ModelInfo object with model details

        Raises:
            FileNotFoundError: If model file doesn't exist
            ValueError: If model format is invalid
        """
        import time

        if not os.path.exists(path):
            raise FileNotFoundError(f"Model file not found: {path}")

        start_time = time.perf_counter_ns()

        # Get file size
        file_size = os.path.getsize(path)

        # Mock model loading - in production this would call native library
        self._model = {
            "path": path,
            "loaded": True,
        }

        self._model_info = ModelInfo(
            name=os.path.basename(path).replace(".gguf", ""),
            path=path,
            architecture="llama",  # Would be detected from GGUF
            parameters=0,  # Would be read from GGUF metadata
            context_length=self._config.max_context_length,
            vocab_size=32000,  # Would be read from GGUF metadata
            quantization="Q4_0",  # Would be read from GGUF metadata
            file_size_bytes=file_size,
        )

        self._stats.load_time_ns = time.perf_counter_ns() - start_time
        return self._model_info

    def unload_model(self) -> None:
        """Unload the current model and free resources."""
        self._model = None
        self._model_info = None
        self._stats = InferenceStats()

    def generate(
        self,
        prompt: str,
        max_tokens: Optional[int] = None,
        temperature: Optional[float] = None,
        top_p: Optional[float] = None,
        top_k: Optional[int] = None,
        stop_sequences: Optional[List[str]] = None,
    ) -> str:
        """
        Generate text from a prompt.

        Args:
            prompt: Input prompt text
            max_tokens: Maximum tokens to generate (overrides config)
            temperature: Sampling temperature (overrides config)
            top_p: Top-p sampling threshold (overrides config)
            top_k: Top-k sampling (overrides config)
            stop_sequences: List of sequences that stop generation

        Returns:
            Generated text

        Raises:
            RuntimeError: If no model is loaded
        """
        if not self.is_loaded:
            raise RuntimeError("No model loaded. Call load_model() first.")

        # Use provided values or fall back to config
        max_tokens = max_tokens or self._config.max_new_tokens
        temperature = temperature if temperature is not None else self._config.temperature
        top_p = top_p if top_p is not None else self._config.top_p
        top_k = top_k if top_k is not None else self._config.top_k

        import time
        start_time = time.perf_counter_ns()

        # Mock generation - in production this would call native inference
        # Placeholder response for development
        response = f"[Generated response for: {prompt[:50]}...]"

        self._stats.prefill_time_ns = time.perf_counter_ns() - start_time
        self._stats.prompt_tokens = len(prompt.split())
        self._stats.generated_tokens = len(response.split())

        return response

    def generate_streaming(
        self,
        prompt: str,
        max_tokens: Optional[int] = None,
        temperature: Optional[float] = None,
        callback: Optional[Callable[[str], None]] = None,
    ) -> Iterator[str]:
        """
        Generate text with streaming output.

        Args:
            prompt: Input prompt text
            max_tokens: Maximum tokens to generate
            temperature: Sampling temperature
            callback: Optional callback function called for each token

        Yields:
            Generated tokens one at a time

        Example:
            >>> for token in engine.generate_streaming("Hello"):
            ...     print(token, end="", flush=True)
        """
        if not self.is_loaded:
            raise RuntimeError("No model loaded. Call load_model() first.")

        max_tokens = max_tokens or self._config.max_new_tokens

        # Mock streaming generation
        tokens = ["[", "Generated", " ", "response", " ", "for", ":", " ", prompt[:20], "...]"]

        for token in tokens:
            if callback:
                callback(token)
            yield token

    def tokenize(self, text: str) -> List[int]:
        """
        Tokenize text into token IDs.

        Args:
            text: Input text

        Returns:
            List of token IDs
        """
        if not self.is_loaded:
            raise RuntimeError("No model loaded. Call load_model() first.")

        # Mock tokenization - returns approximate token count
        # In production, this would use the actual tokenizer
        words = text.split()
        return list(range(len(words)))

    def detokenize(self, tokens: List[int]) -> str:
        """
        Convert token IDs back to text.

        Args:
            tokens: List of token IDs

        Returns:
            Decoded text
        """
        if not self.is_loaded:
            raise RuntimeError("No model loaded. Call load_model() first.")

        # Mock detokenization
        return f"[Decoded {len(tokens)} tokens]"

    def count_tokens(self, text: str) -> int:
        """
        Count tokens in text without full tokenization.

        Args:
            text: Input text

        Returns:
            Approximate token count
        """
        # Rough approximation: ~4 characters per token
        return max(1, len(text) // 4)

    def chat(
        self,
        messages: List[Dict[str, str]],
        max_tokens: Optional[int] = None,
        temperature: Optional[float] = None,
    ) -> str:
        """
        Generate a chat completion.

        Args:
            messages: List of message dicts with 'role' and 'content'
            max_tokens: Maximum tokens to generate
            temperature: Sampling temperature

        Returns:
            Assistant's response

        Example:
            >>> response = engine.chat([
            ...     {"role": "system", "content": "You are a helpful assistant."},
            ...     {"role": "user", "content": "Hello!"}
            ... ])
        """
        # Format messages into a prompt
        prompt_parts = []
        for msg in messages:
            role = msg.get("role", "user")
            content = msg.get("content", "")
            if role == "system":
                prompt_parts.append(f"System: {content}")
            elif role == "user":
                prompt_parts.append(f"User: {content}")
            elif role == "assistant":
                prompt_parts.append(f"Assistant: {content}")

        prompt_parts.append("Assistant:")
        prompt = "\n".join(prompt_parts)

        return self.generate(prompt, max_tokens=max_tokens, temperature=temperature)


class LlmContext:
    """
    LLM context for framework integration.

    This class wraps the LLM engine to provide a consistent interface
    with other ABI framework modules.

    Example:
        >>> from abi.config import LlmConfig
        >>> ctx = LlmContext(LlmConfig(model_path="./model.gguf"))
        >>> response = ctx.generate("Hello!")
    """

    def __init__(self, config: Optional["LlmConfig"] = None):
        """
        Initialize the LLM context.

        Args:
            config: LLM configuration from abi.config
        """
        from .config import LlmConfig

        self._config = config or LlmConfig.defaults()
        self._engine: Optional[LlmEngine] = None

    @property
    def engine(self) -> LlmEngine:
        """Get or create the LLM engine."""
        if self._engine is None:
            inference_config = InferenceConfig(
                max_context_length=self._config.context_size,
                batch_size=self._config.batch_size,
                use_gpu=self._config.use_gpu,
                num_threads=self._config.threads or 0,
            )
            self._engine = LlmEngine(inference_config)

            # Auto-load model if path is provided
            if self._config.model_path:
                self._engine.load_model(self._config.model_path)

        return self._engine

    def generate(self, prompt: str, **kwargs) -> str:
        """Generate text from prompt."""
        return self.engine.generate(prompt, **kwargs)

    def tokenize(self, text: str) -> List[int]:
        """Tokenize text."""
        return self.engine.tokenize(text)

    def detokenize(self, tokens: List[int]) -> str:
        """Detokenize tokens."""
        return self.engine.detokenize(tokens)

    def close(self) -> None:
        """Close and release resources."""
        if self._engine:
            self._engine.unload_model()
            self._engine = None


# Convenience functions

def infer(model_path: str, prompt: str, **kwargs) -> str:
    """
    Quick inference helper.

    Args:
        model_path: Path to GGUF model file
        prompt: Input prompt
        **kwargs: Additional inference parameters

    Returns:
        Generated text

    Example:
        >>> result = infer("./model.gguf", "Hello, world!")
    """
    engine = LlmEngine()
    engine.load_model(model_path)
    return engine.generate(prompt, **kwargs)


def list_models(directory: str = "./models") -> List[str]:
    """
    List available GGUF models in a directory.

    Args:
        directory: Directory to search for models

    Returns:
        List of model file paths
    """
    models = []
    if os.path.exists(directory):
        for filename in os.listdir(directory):
            if filename.endswith(".gguf"):
                models.append(os.path.join(directory, filename))
    return models


def is_enabled() -> bool:
    """Check if LLM features are available."""
    # In mock mode, always return True
    # In production, this would check native library availability
    return True
